cpp/src/buffer/buffer.cpp

```diff
@@ -61,175 +61,69 @@ std::byte const* Buffer::data() const {
+        "src_offset + size can't be greater than src.size",
         std::invalid_argument
     );
-
-    if (size == 0) {  // empty buffer, nothing to do
-        return 0;
+    if (size == 0) {
+        return;  // Nothing to copy.
     }
 
-    // lambda to perform a memcpy and optionally attach an event
-    auto do_cuda_memcpy_async = [&](cudaMemcpyKind kind, cuda::std::uint8_t* source) {
-        // if this buffer has an event, ask the current stream to wait for it, before
-        // performing the memcpy
-        if (event_) {
-            event_->stream_wait(stream);
-        }
-
-        RAPIDSMPF_CUDA_TRY(
-            cudaMemcpyAsync(dest.data() + dest_offset, source, size, kind, stream)
-        );
-
-        // if the copy is not host-to-host, override the event to track the async copy
-        if (attach_event && kind != cudaMemcpyHostToHost) {
-            dest.override_event(CudaEvent::make_shared_record(stream));
-        }
-
-        return std::ptrdiff_t(size);
-    };
-
-    return std::visit(
-        overloaded{
-            [&](const HostStorageT& storage) {
-                switch (dest.mem_type()) {
-                case MemoryType::HOST:  // host -> host
-                    return do_cuda_memcpy_async(cudaMemcpyHostToHost, storage->data());
-                case MemoryType::DEVICE:  // host -> device
-                    return do_cuda_memcpy_async(cudaMemcpyHostToDevice, storage->data());
-                }
-                RAPIDSMPF_FAIL("Invalid memory type");  // unreachable
-            },
-            [&](const DeviceStorageT& storage) {
-                switch (dest.mem_type()) {
-                case MemoryType::HOST:  // device -> host
-                    return do_cuda_memcpy_async(
-                        cudaMemcpyDeviceToHost,
-                        static_cast<cuda::std::uint8_t*>(storage->data())
-                    );
-                case MemoryType::DEVICE:  // device -> device
-                    return do_cuda_memcpy_async(
-                        cudaMemcpyDeviceToDevice,
-                        static_cast<cuda::std::uint8_t*>(storage->data())
-                    );
-                }
-                RAPIDSMPF_FAIL("Invalid memory type");  // unreachable
-            }
-        },
-        storage_
-    );
-}
+    // Make sure we wait on any buffer event.
+    if (auto e = dst.get_event()) {
+        e->stream_wait(stream);
+    }
+    if (auto e = src.get_event()) {
+        e->stream_wait(stream);
+    }
 
-bool Buffer::is_ready() const {
-    return !event_ || event_->is_ready();
-}
+    RAPIDSMPF_CUDA_TRY(cudaMemcpyAsync(
+        dst.data() + dst_offset, src.data() + src_offset, size, cudaMemcpyDefault, stream
+    ));
 
-void Buffer::wait_for_ready() const {
-    if (event_) {
-        event_->host_wait();
+    // Override the event to track the async copy.
+    if (attach_cuda_event) {
+        dst.override_event(CudaEvent::make_shared_record(stream));
     }
```

Comment:
Do we not need to override the event to the `src` as well? I _think_ it's unsafe not to, for example, if `cudaFreeAsync(src, src_stream)`, where `src_stream` is a different stream than the `stream` passed to `buffer_copy`, this will be potentially UB.

Reply:
good catch